js/zb.js

```diff
@@ -1474,15 +1474,22 @@ module.exports = class zb extends Exchange {
         request[periodField] = timeframeValue;
         const sizeField = swap ? 'size' : 'limit';
         request[sizeField] = limit;
-        const method = this.getSupportedMapping (market['type'], {
+        const price = this.safeString (params, 'price');
+        params = this.omit (params, 'price');
+        let method = this.getSupportedMapping (market['type'], {
             'spot': 'spotV1PublicGetKline',
             'swap': 'contractV1PublicGetKline',
         });
-        if (since !== undefined) {
-            request['since'] = since;
-        }
-        if (limit !== undefined) {
-            request['size'] = limit;
+        if (market['swap']) {
+            if (price === 'mark') {
+                method = 'contractV1PublicGetMarkKline';
```

Comment:
Have you tried these endpoints with a `since` param? Cause in order to get the history, the user must have some means of pagination.

Reply:
I excluded it because the documentation for swaps says there is no since param, but there is for spot https://github.com/ZBFuture/docs_en/blob/main/API%20V2%20_en.md#73--candlestick
They can still adjust the limit and it returns the most recent values

Reply:
I just tested it and the swap endpoints work with the since param

Reply:
@Dan-krm i would suggest trying two params `startTime` (seconds and milliseconds) and `since` (also seconds and milliseconds)  – just to see if anything of that works and if we can use date-based pagination with those endpoints.

Reply:
> I just tested it and the swap endpoints work with the since param

And the returned datetimes are properly aligned according to your `since` setting? Or not?

```
node examples/js/cli zb fetchIndexOHLCV BTC/USDT 1m 2022-03-01T00:00:00Z --iso8601 --table
```

Reply:
What I tested to see if it worked was this:
```
zb.fetchMarkOHLCV (BTC/USDT:USDT, 1m, 1648159620000)
2022-03-24T22:39:26.435Z iteration 0 passed in 577 ms

1648159620000 | 43807.17 | 43810.59 | 43806.19 | 43807.73 | 
1648159680000 | 43807.09 | 43807.09 | 43779.86 | 43781.52 | 
1648159740000 | 43782.67 | 43790.81 | 43779.97 | 43786.96 | 
1648159800000 | 43786.96 | 43789.27 | 43785.34 | 43786.28 | 
1648159860000 |  43786.3 | 43799.03 | 43785.91 | 43794.52 | 
1648159920000 | 43794.92 |  43798.7 | 43790.38 | 43798.65 | 
1648159980000 | 43798.68 | 43799.53 | 43790.13 | 43792.14 | 
1648160040000 | 43792.05 | 43818.26 | 43790.22 | 43793.68 | 
1648160100000 | 43792.76 | 43794.56 | 43783.57 | 43785.91 | 
1648160160000 | 43785.91 | 43787.85 | 43784.93 | 43785.51 | 
1648160220000 | 43786.99 | 43786.99 | 43785.03 | 43785.04 | 
1648160280000 | 43785.04 | 43788.26 | 43785.04 | 43788.26 | 
1648160340000 | 43785.74 | 43793.24 | 43785.31 | 43790.97 | 
1648160400000 |  43790.5 | 43799.08 | 43789.27 | 43789.27 | 
1648160460000 | 43793.27 |  43793.8 | 43781.28 | 43785.87 | 
1648160520000 | 43789.24 | 43812.89 | 43789.24 | 43807.49 | 
1648160580000 | 43807.48 | 43808.06 | 43784.77 | 43799.68 | 
1648160640000 | 43799.67 | 43807.04 | 43794.28 | 43807.04 | 
1648160700000 | 43807.03 | 43811.52 | 43801.16 | 43801.16 | 
1648160760000 | 43801.15 | 43835.08 | 43801.15 | 43831.71 | 
1648160820000 | 43831.69 | 43859.72 | 43826.97 | 43859.72 | 
1648160880000 |  43859.7 |    43890 | 43851.88 | 43878.73 | 
1648160940000 | 43878.71 | 43884.28 | 43829.23 | 43848.78 | 
1648161000000 | 43848.77 | 43852.49 |  43827.5 | 43827.73 | 
1648161060000 | 43827.29 | 43831.81 | 43826.53 | 43826.53 | 
1648161120000 | 43826.53 |  43847.1 | 43826.53 |  43847.1 | 
1648161180000 |  43846.8 |  43850.9 |  43846.8 | 43850.51 | 
1648161240000 | 43849.16 | 43850.58 | 43847.39 | 43850.48 | 
1648161300000 | 43849.16 | 43857.12 | 43847.06 | 43854.39 | 
1648161360000 | 43854.93 | 43864.06 | 43846.31 | 43847.97 | 
1648161420000 | 43848.13 | 43848.13 |  43830.7 | 43833.32 | 
1648161480000 | 43836.45 | 43838.33 | 43834.53 | 43835.64 | 
1648161540000 | 43834.53 | 43835.74 | 43834.33 | 43834.33 | 
33 objects
```

Reply:
> And the returned datetimes are properly aligned according to your `since` setting? Or not?

If I try the request you mentioned it only returns the most recent values and the max size is 1440 values

Reply:
How does your verbose request look like with params embedded ? Have you tried startTime ?

Reply:
> What I tested to see if it worked was this:

We want to not only test if it works (it will work with any arbitrary params added to the URL, cause it's a public GET request, and public GET requests don't care about the excessive params). We also want to see if it returns the correct results as was requested or not.

Reply:
startTime and since seem to work the same way and just return the most recent values with the max amount being 1440 values, trying to set and handle endTime didn't work for me. Using both milliseconds and seconds in the since argument still returned the most recent values
Since:
```
node examples/js/cli zb fetchMarkOHLCV BTC/USDT:USDT 15m 2022-03-10T12:00:00Z 10 --iso8601 --table --verbose
2022-03-24T22:47:49.425Z
Node.js: v16.9.1
CCXT v1.77.6
zb.fetchMarkOHLCV (BTC/USDT:USDT, 15m, 1646913600000, 10)
fetch Request:
 zb GET https://fapi.zb.com/api/public/v1/markKline?symbol=BTC_USDT&period=15M&size=10&since=1646913600000
RequestHeaders:
 {}
RequestBody:
 undefined

handleRestResponse:
 zb GET https://fapi.zb.com/api/public/v1/markKline?symbol=BTC_USDT&period=15M&size=10&since=1646913600000 200 OK
ResponseHeaders:
 {
  'Access-Control-Allow-Credentials': 'true',
  Connection: 'keep-alive',
  'Content-Length': '521',
  'Content-Type': 'application/json',
  Date: 'Thu, 24 Mar 2022 22:47:53 GMT',
  Server: 'nginx'
}
ResponseBody:
 {"code":10000,"desc":"操作成功","data":[[43774.95,43818.67,43722.62,43749.8,1648153800],[43749.77,43759.12,43700,43704.53,1648154700],[43705.83,43723.16,43700,43702.58,1648155600],[43700.99,43770.81,43700,43762.49,1648156500],[43762.46,43805.56,43725.34,43743.39,1648157400],[43742.18,43845.52,43738.07,43817.43,1648158300],[43817.42,43829.89,43779.86,43793.68,1648159200],[43792.76,43890,43781.28,43848.78,1648160100],[43848.77,43864.06,43826.53,43857.43,1648161000],[43857.99,43869.4,43836.01,43839.28,1648161900]]}

2022-03-24T22:47:53.097Z iteration 0 passed in 363 ms

2022-03-24T20:30:00.000Z | 43774.95 | 43818.67 | 43722.62 |  43749.8 | 
2022-03-24T20:45:00.000Z | 43749.77 | 43759.12 |    43700 | 43704.53 | 
2022-03-24T21:00:00.000Z | 43705.83 | 43723.16 |    43700 | 43702.58 | 
2022-03-24T21:15:00.000Z | 43700.99 | 43770.81 |    43700 | 43762.49 | 
2022-03-24T21:30:00.000Z | 43762.46 | 43805.56 | 43725.34 | 43743.39 | 
2022-03-24T21:45:00.000Z | 43742.18 | 43845.52 | 43738.07 | 43817.43 | 
2022-03-24T22:00:00.000Z | 43817.42 | 43829.89 | 43779.86 | 43793.68 | 
2022-03-24T22:15:00.000Z | 43792.76 |    43890 | 43781.28 | 43848.78 | 
2022-03-24T22:30:00.000Z | 43848.77 | 43864.06 | 43826.53 | 43857.43 | 
2022-03-24T22:45:00.000Z | 43857.99 |  43869.4 | 43836.01 | 43839.28 | 
10 objects
```
startTime:
```
node examples/js/cli zb fetchMarkOHLCV BTC/USDT:USDT 15m 2022-03-10T12:00:00Z 10 --iso8601 --table --verbose
2022-03-24T22:51:54.835Z
Node.js: v16.9.1
CCXT v1.77.6
zb.fetchMarkOHLCV (BTC/USDT:USDT, 15m, 1646913600000, 10)
fetch Request:
 zb GET https://fapi.zb.com/api/public/v1/markKline?symbol=BTC_USDT&period=15M&size=10&startTime=1646913600000
RequestHeaders:
 {}
RequestBody:
 undefined

handleRestResponse:
 zb GET https://fapi.zb.com/api/public/v1/markKline?symbol=BTC_USDT&period=15M&size=10&startTime=1646913600000 200 OK
ResponseHeaders:
 {
  'Access-Control-Allow-Credentials': 'true',
  Connection: 'keep-alive',
  'Content-Length': '521',
  'Content-Type': 'application/json',
  Date: 'Thu, 24 Mar 2022 22:51:59 GMT',
  Server: 'nginx'
}
ResponseBody:
 {"code":10000,"desc":"操作成功","data":[[43774.95,43818.67,43722.62,43749.8,1648153800],[43749.77,43759.12,43700,43704.53,1648154700],[43705.83,43723.16,43700,43702.58,1648155600],[43700.99,43770.81,43700,43762.49,1648156500],[43762.46,43805.56,43725.34,43743.39,1648157400],[43742.18,43845.52,43738.07,43817.43,1648158300],[43817.42,43829.89,43779.86,43793.68,1648159200],[43792.76,43890,43781.28,43848.78,1648160100],[43848.77,43864.06,43826.53,43857.43,1648161000],[43857.99,43869.4,43836.01,43839.27,1648161900]]}

2022-03-24T22:51:58.434Z iteration 0 passed in 371 ms

2022-03-24T20:30:00.000Z | 43774.95 | 43818.67 | 43722.62 |  43749.8 | 
2022-03-24T20:45:00.000Z | 43749.77 | 43759.12 |    43700 | 43704.53 | 
2022-03-24T21:00:00.000Z | 43705.83 | 43723.16 |    43700 | 43702.58 | 
2022-03-24T21:15:00.000Z | 43700.99 | 43770.81 |    43700 | 43762.49 | 
2022-03-24T21:30:00.000Z | 43762.46 | 43805.56 | 43725.34 | 43743.39 | 
2022-03-24T21:45:00.000Z | 43742.18 | 43845.52 | 43738.07 | 43817.43 | 
2022-03-24T22:00:00.000Z | 43817.42 | 43829.89 | 43779.86 | 43793.68 | 
2022-03-24T22:15:00.000Z | 43792.76 |    43890 | 43781.28 | 43848.78 | 
2022-03-24T22:30:00.000Z | 43848.77 | 43864.06 | 43826.53 | 43857.43 | 
2022-03-24T22:45:00.000Z | 43857.99 |  43869.4 | 43836.01 | 43839.27 | 
10 objects
```

Reply:
Ok, then we should not send neither `since` nor `startTime` for swaps.

Reply:
> Ok, then we should not send neither `since` nor `startTime` for swaps.

Okay it seems like since behaves the same way for spot markets as well should I just remove since altogether?

Reply:
I would keep `since` for spot markets, cause it's documented (despite that it does not work):

Milliseconds:

- https://api.zb.com/data/v1/kline?market=btc_usdt&type=15min&limit=1000&since=1646092800000&size=1000

Seconds:

- https://api.zb.com/data/v1/kline?market=btc_usdt&type=15min&limit=1000&since=1646092800&size=1000

Also, we should not send both `limit` and `size`, it's either this or that depending on the market type.

Reply:
Looks like it must be `size` for both swaps and spot markets. The `limit` does not belong there.